--- a/drivers/remoteproc/remoteproc_virtio.c
+++ b/drivers/remoteproc/remoteproc_virtio.c
@@ -333,6 +333,9 @@ int rproc_add_virtio_dev(struct rproc_vdev *rvdev, int id)
 	 */
 	get_device(&rproc->dev);
 
+	/* store cpu id of remote proc to pass it to rpmsg framework */
+	vdev->priv = &rproc->cpu;
+
 	ret = register_virtio_device(vdev);
 	if (ret) {
 		put_device(&rproc->dev);

--- a/include/linux/remoteproc.h
+++ b/include/linux/remoteproc.h
@@ -405,6 +405,7 @@ enum rproc_crash_type {
  * @cached_table: copy of the resource table
  * @table_csum: checksum of the resource table
  * @has_iommu: flag to indicate if remote processor is behind an MMU
+ * @cpu: cpu id for remote firmware
  */
 struct rproc {
 	struct klist_node node;
@@ -436,6 +437,7 @@ struct rproc {
 	struct resource_table *cached_table;
 	u32 table_csum;
	bool has_iommu;
+	u32 cpu;
 };
 
 /* we currently support only two vrings per rvdev */
