From e3da44dc939b29256f2c70940ea79a80082f6540 Mon Sep 17 00:00:00 2001
From: Ahsan Hussain <ahsan_hussain@mentor.com>
Date: Thu, 12 Apr 2018 16:25:16 +0500
Subject: [PATCH 3/3] enable building the added lttng modules

Signed-off-by: Ahsan Hussain <ahsan_hussain@mentor.com>
---
 probes/Kbuild | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/probes/Kbuild b/probes/Kbuild
index d5d159a..1b9276a 100644
--- a/probes/Kbuild
+++ b/probes/Kbuild
@@ -261,4 +261,9 @@ ifneq ($(CONFIG_DYNAMIC_FTRACE),)
   endif
 endif # CONFIG_DYNAMIC_FTRACE
 
+ifneq ($(CONFIG_MEMF),)
+    obj-$(CONFIG_LTTNG) += lttng-probe-memf.o
+    obj-$(CONFIG_LTTNG) += lttng-probe-Synchronization.o
+endif # CONFIG_MEMF
+
 # vim:syntax=make
-- 
2.8.1

