diff --git c/configure i/configure
index cb8d78f..862306c 100755
--- c/configure
+++ i/configure
@@ -735,6 +735,7 @@ OPT_MAC_SDK=
 COMMERCIAL_USER=ask
 LICENSE_FILE=
 CFG_DEV=no
+CFG_SOURCERYANALYZER=no
 
 # initalize variables used for installation
 QT_INSTALL_PREFIX=
@@ -2158,6 +2159,13 @@ while [ "$#" -gt 0 ]; do
             UNKNOWN_OPT=yes
         fi
         ;;
+    sa-trace)
+        if [ "$VAL" = "yes" ]; then
+            CFG_SOURCERYANALYZER="yes"
+        else
+            UNKNOWN_OPT=yes
+        fi
+        ;;
     *)
         UNKNOWN_OPT=yes
         ;;
@@ -5681,7 +5689,9 @@ if [ "$CFG_WIDGETS" = "no" ]; then
     QT_CONFIG="$QT_CONFIG no-widgets"
     QCONFIG_FLAGS="$QCONFIG_FLAGS QT_NO_WIDGETS"
 fi
-
+if [ "$CFG_SOURCERYANALYZER" = "yes" ]; then
+    QT_CONFIG="$QT_CONFIG tracesupport"
+fi
 if [ "$XPLATFORM_MAC" = "yes" ]; then
     #On Mac we implicitly link against libz, so we
     #never use the 3rdparty stuff.
@@ -6568,6 +6578,7 @@ report_support "    TDS .................." "$CFG_SQL_tds" plugin "plugin" yes "
 report_support "  udev ..................." "$CFG_LIBUDEV"
 report_support "  xkbcommon .............." "$CFG_XKBCOMMON" system "system library" qt "bundled copy, XKB config root: $CFG_XKB_CONFIG_ROOT"
 report_support "  zlib ..................." "$CFG_ZLIB" system "system library" yes "bundled copy"
+report_support "  Sourcery Analyzer ......" "$CFG_SOURCERYANALYZER"
 
 echo
 
diff --git c/mkspecs/features/tracesupport.prf i/mkspecs/features/tracesupport.prf
new file mode 100644
index 0000000..d94a437
--- /dev/null
+++ i/mkspecs/features/tracesupport.prf
@@ -0,0 +1,3 @@
+DEFINES += ENABLE_SA_TRACE
+CFLAGS          += -g3 -O1
+LIBS            += -L$$[QT_INSTALL_LIBS] -lQt5TraceSupport -ldl -llttng-ust -lurcu-bp
diff --git c/src/corelib/animation/qabstractanimation.cpp i/src/corelib/animation/qabstractanimation.cpp
index 95d7713..a539807 100644
--- c/src/corelib/animation/qabstractanimation.cpp
+++ i/src/corelib/animation/qabstractanimation.cpp
@@ -155,6 +155,10 @@
 #include <QtCore/qcoreevent.h>
 #include <QtCore/qpointer.h>
 
+#ifdef ENABLE_SA_TRACE
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 #ifndef QT_NO_ANIMATION
 
 #define DEFAULT_TIMER_INTERVAL 16
@@ -224,6 +228,10 @@ QUnifiedTimer::QUnifiedTimer() :
     startTimersPending(false), stopTimerPending(false),
     slowdownFactor(5.0f), profilerCallback(0)
 {
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtAnimTimingInterval, DEFAULT_TIMER_INTERVAL);
+#endif // ENABLE_SA_TRACE
+
     time.invalidate();
     driver = &defaultDriver;
 }
@@ -341,6 +349,10 @@ void QUnifiedTimer::restart()
 
 void QUnifiedTimer::setTimingInterval(int interval)
 {
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtAnimTimingInterval, interval);
+#endif // ENABLE_SA_TRACE
+
     timingInterval = interval;
 
     if (driver->isRunning() && !pauseTimer.isActive()) {
@@ -572,6 +584,10 @@ void QAnimationTimer::updateAnimationsTime(qint64 delta)
     //it might happen in some cases that the time doesn't change because events are delayed
     //when the CPU load is high
     if (delta) {
+#ifdef ENABLE_SA_TRACE
+        qt_tracepoint(Qt, qtAnimTick, 1);
+#endif // ENABLE_SA_TRACE
+
         insideTick = true;
         for (currentAnimationIdx = 0; currentAnimationIdx < animations.count(); ++currentAnimationIdx) {
             QAbstractAnimation *animation = animations.at(currentAnimationIdx);
@@ -581,6 +597,10 @@ void QAnimationTimer::updateAnimationsTime(qint64 delta)
         }
         insideTick = false;
         currentAnimationIdx = 0;
+        
+#ifdef ENABLE_SA_TRACE
+        qt_tracepoint(Qt, qtAnimTick, 0);
+#endif // ENABLE_SA_TRACE
     }
 }
 
@@ -828,6 +848,10 @@ bool QAnimationDriver::isRunning() const
 
 void QAnimationDriver::start()
 {
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtAnimDriver, 1);
+#endif // ENABLE_SA_TRACE
+
     Q_D(QAnimationDriver);
     if (!d->running) {
         d->running = true;
@@ -843,6 +867,10 @@ void QAnimationDriver::stop()
         d->running = false;
         emit stopped();
     }
+
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtAnimDriver, 0);
+#endif // ENABLE_SA_TRACE
 }
 
 
diff --git c/src/corelib/corelib.pro i/src/corelib/corelib.pro
index 812aee3..c116a3c 100644
--- c/src/corelib/corelib.pro
+++ i/src/corelib/corelib.pro
@@ -32,6 +32,9 @@ ANDROID_PERMISSIONS = \
 
 load(qt_module)
 
+contains(QT_CONFIG,tracesupport):CONFIG += tracesupport
+contains(QT_CONFIG,tracesupport):LIBS += -L$$DESTDIR
+
 include(animation/animation.pri)
 include(arch/arch.pri)
 include(global/global.pri)
diff --git c/src/corelib/kernel/qcoreapplication.cpp i/src/corelib/kernel/qcoreapplication.cpp
index 6868eb6..52f9f0d 100644
--- c/src/corelib/kernel/qcoreapplication.cpp
+++ i/src/corelib/kernel/qcoreapplication.cpp
@@ -112,6 +112,11 @@
 
 #include <algorithm>
 
+#ifdef ENABLE_SA_TRACE
+    #define QT_TRACEPOINT_PER_LIBRARY_DEFINITION
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 QT_BEGIN_NAMESPACE
 
 #ifndef QT_NO_QOBJECT
@@ -406,6 +411,10 @@ QCoreApplicationPrivate::QCoreApplicationPrivate(int &aargc, char **aargv, uint
     , q_ptr(0)
 #endif
 {
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtApplicationCreated, 1);
+#endif // ENABLE_SA_TRACE
+
     app_compile_version = flags & 0xffffff;
     static const char *const empty = "";
     if (argc == 0 || argv == 0) {
@@ -443,6 +452,10 @@ QCoreApplicationPrivate::~QCoreApplicationPrivate()
     delete [] origArgv;
 #endif
     QCoreApplicationPrivate::clearApplicationFilePath();
+
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtApplicationCreated, 0);
+#endif // ENABLE_SA_TRACE
 }
 
 #ifndef QT_NO_QOBJECT
@@ -917,6 +930,11 @@ void QCoreApplication::setQuitLockEnabled(bool enabled)
 */
 bool QCoreApplication::notifyInternal(QObject *receiver, QEvent *event)
 {
+#ifdef ENABLE_SA_TRACE
+    if (event)
+        qt_tracepoint(Qt, qtCoreEvents, event->type());
+#endif // ENABLE_SA_TRACE
+
     // Make it possible for Qt Script to hook into events even
     // though QApplication is subclassed...
     bool result = false;
diff --git c/src/gui/gui.pro i/src/gui/gui.pro
index af84eb7..f9c8728 100644
--- c/src/gui/gui.pro
+++ i/src/gui/gui.pro
@@ -46,6 +46,8 @@ include(opengl/opengl.pri)
 include(animation/animation.pri)
 include(itemmodels/itemmodels.pri)
 
+contains(QT_CONFIG,tracesupport):CONFIG += tracesupport
+
 QMAKE_LIBS += $$QMAKE_LIBS_GUI
 
 load(cmake_functions)
diff --git c/src/gui/image/qimagereader.cpp i/src/gui/image/qimagereader.cpp
index a10198a..331e972 100644
--- c/src/gui/image/qimagereader.cpp
+++ i/src/gui/image/qimagereader.cpp
@@ -152,6 +152,11 @@
 
 #include <algorithm>
 
+#ifdef ENABLE_SA_TRACE
+    #include <QThread>
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 QT_BEGIN_NAMESPACE
 
 #ifndef QT_NO_IMAGEFORMATPLUGIN
@@ -1172,6 +1177,11 @@ bool QImageReader::read(QImage *image)
     if (!d->handler && !d->initHandler())
         return false;
 
+#ifdef ENABLE_SA_TRACE
+    long threadId = (long)(QThread::currentThread());
+    qt_tracepoint(Qt, qtImageReader, threadId, 1, fileName().toLatin1().constData());
+#endif // ENABLE_SA_TRACE
+
     // set the handler specific options.
     if (d->handler->supportsOption(QImageIOHandler::ScaledSize) && d->scaledSize.isValid()) {
         if ((d->handler->supportsOption(QImageIOHandler::ClipRect) && !d->clipRect.isNull())
@@ -1192,6 +1202,9 @@ bool QImageReader::read(QImage *image)
     if (!d->handler->read(image)) {
         d->imageReaderError = InvalidDataError;
         d->errorString = QLatin1String(QT_TRANSLATE_NOOP(QImageReader, "Unable to read image data"));
+#ifdef ENABLE_SA_TRACE
+        qt_tracepoint(Qt, qtImageReader, threadId, 0, fileName().toLatin1().constData());
+#endif // ENABLE_SA_TRACE
         return false;
     }
 
@@ -1252,6 +1265,10 @@ bool QImageReader::read(QImage *image)
            image->setDevicePixelRatio(2.0);
     }
 
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtImageReader, threadId, 0, fileName().toLatin1().constData());
+#endif // ENABLE_SA_TRACE
+
     return true;
 }
 
diff --git c/src/gui/kernel/qguiapplication.cpp i/src/gui/kernel/qguiapplication.cpp
index 7648233..5716cca 100644
--- c/src/gui/kernel/qguiapplication.cpp
+++ i/src/gui/kernel/qguiapplication.cpp
@@ -109,6 +109,11 @@
 
 #include <ctype.h>
 
+#ifdef ENABLE_SA_TRACE
+    #define QT_TRACEPOINT_PER_LIBRARY_DEFINITION
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 QT_BEGIN_NAMESPACE
 
 Q_GUI_EXPORT bool qt_is_gui_used = true;
@@ -1487,6 +1492,35 @@ int QGuiApplication::exec()
 */
 bool QGuiApplication::notify(QObject *object, QEvent *event)
 {
+#ifdef ENABLE_SA_TRACE
+    if (event) {
+        switch (event->type())
+        {
+        case QEvent::KeyPress:
+        case QEvent::KeyRelease:
+            {
+                QKeyEvent* key = static_cast<QKeyEvent*>(event);
+                qt_tracepoint(Qt, qtNotifyKeyEvent, key->type(), key->key(), key->modifiers());
+            }
+            break;
+
+        case QEvent::MouseButtonPress:
+        case QEvent::MouseButtonRelease:
+        case QEvent::MouseButtonDblClick:
+        case QEvent::MouseMove:
+            {
+                QMouseEvent* mouse = static_cast<QMouseEvent*>(event);
+                qt_tracepoint(Qt, qtSendMouseEvent, mouse->buttons());
+            }
+            break;
+
+        default:
+            // Ignore
+            break;
+        }
+    }
+#endif // ENABLE_SA_TRACE
+
 #ifndef QT_NO_SHORTCUT
     if (event->type() == QEvent::KeyPress) {
         // Try looking for a Shortcut before sending key events
diff --git c/src/src.pro i/src/src.pro
index 6a805a6..3ae220b 100644
--- c/src/src.pro
+++ i/src/src.pro
@@ -117,7 +117,11 @@ src_plugins.depends = src_sql src_xml src_network
 
 src_android.subdir = $$PWD/android
 
+src_tracesupport.subdir = $$PWD/tracesupport
+src_tracesupport.target = sub-tracesupport
+
 # this order is important
+contains(QT_CONFIG, tracesupport): SUBDIRS += src_tracesupport
 SUBDIRS += src_tools_bootstrap src_tools_moc src_tools_rcc src_corelib src_tools_qlalr
 TOOLS = src_tools_moc src_tools_rcc src_tools_qlalr
 win32:SUBDIRS += src_winmain
diff --git c/src/tracesupport/qt_tracepoints.c i/src/tracesupport/qt_tracepoints.c
new file mode 100644
index 0000000..121df3d
--- /dev/null
+++ i/src/tracesupport/qt_tracepoints.c
@@ -0,0 +1,2 @@
+#define TRACEPOINT_CREATE_PROBES
+#include "qt_tracepoints.h"
diff --git c/src/tracesupport/qt_tracepoints.h i/src/tracesupport/qt_tracepoints.h
new file mode 100644
index 0000000..f1e33ba
--- /dev/null
+++ i/src/tracesupport/qt_tracepoints.h
@@ -0,0 +1,85 @@
+#undef TRACEPOINT_PROVIDER
+#define TRACEPOINT_PROVIDER         Qt
+
+#undef TRACEPOINT_INCLUDE_FILE
+#define TRACEPOINT_INCLUDE_FILE     ./qt_tracepoints.h
+
+#if !defined(_QT_TRACEPOINTS_H) || defined(TRACEPOINT_HEADER_MULTI_READ)
+#define _QT_TRACEPOINTS_H
+
+/* This define will be present exactly once per output library */
+#ifdef QT_TRACEPOINT_PER_LIBRARY_DEFINITION
+    #define TRACEPOINT_DEFINE
+    #define TRACEPOINT_PROBE_DYNAMIC_LINKAGE
+#endif
+
+#include <lttng/tracepoint.h>
+#include <stdint.h>
+
+/* Map qt_tracepont to the tracepoint implementation used: */
+#define qt_tracepoint tracepoint
+
+/* Startup Agent, app start stop. */
+TRACEPOINT_EVENT(Qt, qtApplicationCreated,
+                TP_ARGS(int, isAlive),
+                TP_FIELDS(ctf_integer(int, isAlive, isAlive)))
+
+/* Qt event processed by Qt core event loop (event types & name) */
+TRACEPOINT_EVENT(Qt, qtQmlResourceLoad,
+                 TP_ARGS(int, isLoading, const char *, qmlFileName),
+                 TP_FIELDS(ctf_integer(int, isLoading, isLoading)
+                           ctf_string(qmlFileName, qmlFileName)))
+
+/* To show the start & end of Qt QML processing events (1=> Start, 0=> Stop) */
+TRACEPOINT_EVENT(Qt, qtQmlFrameCompile,
+                 TP_ARGS(int, isCompiling, const char *, qmlFileName),
+                 TP_FIELDS(ctf_integer(int, isCompiling, isCompiling) ctf_string(qmlFileName, qmlFileName)))
+TRACEPOINT_EVENT(Qt, qtQmlFrameCreate,
+                 TP_ARGS(int, isCreating, const char *, qmlFileName),
+                 TP_FIELDS(ctf_integer(int, isCreating,  isCreating) ctf_string(qmlFileName, qmlFileName)))
+TRACEPOINT_EVENT(Qt, qtQmlFrameBind,
+                 TP_ARGS(int, isBinding, const char *, qmlFileName),
+                 TP_FIELDS(ctf_integer(int, isBinding,   isBinding) ctf_string(qmlFileName, qmlFileName)))
+
+/* Qt Frame Sync (OpenGL or Raster) */
+TRACEPOINT_EVENT(Qt, qtFrameSync, TP_ARGS(int, isRendering),  TP_FIELDS(ctf_integer(int, isRendering, isRendering)))
+
+/* Qt animation framework timer activity */
+TRACEPOINT_EVENT(Qt, qtAnimDriver, TP_ARGS(int, isActive), TP_FIELDS(ctf_integer(int, isActive, isActive)))
+TRACEPOINT_EVENT(Qt, qtAnimTick, TP_ARGS(int, isUpdating), TP_FIELDS(ctf_integer(int, isUpdating, isUpdating)))
+TRACEPOINT_EVENT(Qt, qtAnimTimingInterval, TP_ARGS(int, interval), TP_FIELDS(ctf_integer(int, interval, interval)))
+
+/* Qt event processed by Qt core event loop (event types & name) */
+TRACEPOINT_EVENT(Qt, qtCoreEvents,
+                 TP_ARGS(int, eventType),
+                 TP_FIELDS(ctf_integer(int, eventType, eventType)))
+
+/* Qt event processed by Qt core event loop (event types & name) */
+TRACEPOINT_EVENT(Qt, qtGraphicsSceneEvents,
+                 TP_ARGS(int, eventType),
+                 TP_FIELDS(ctf_integer(int, eventType, eventType)))
+
+/* Qt QSendMouseEvent (event types & name) */
+TRACEPOINT_EVENT(Qt, qtSendMouseEvent, TP_ARGS(int, buttons), TP_FIELDS(ctf_integer(int, buttons, buttons)))
+TRACEPOINT_EVENT(Qt, qtMouseEventConsumed, TP_ARGS(int, buttons), TP_FIELDS(ctf_integer(int, buttons, buttons)))
+
+/* Qt NotifyKeyEvent (event types & name) */
+TRACEPOINT_EVENT(Qt, qtNotifyKeyEvent, TP_ARGS(int, type, int, key, int, modifiers),
+                 TP_FIELDS(ctf_integer(int, type, type)
+                           ctf_integer(int, key, key)
+                           ctf_integer(int, modifiers, modifiers)))
+TRACEPOINT_EVENT(Qt, qtKeyEventConsumed, TP_ARGS(int, type, int, key, int, modifiers),
+                 TP_FIELDS(ctf_integer(int, type, type)
+                           ctf_integer(int, key, key)
+                           ctf_integer(int, modifiers, modifiers)))
+
+/* Qt ImageReader Event (image filename) */
+TRACEPOINT_EVENT(Qt, qtImageReader,
+                 TP_ARGS(int, threadPtr, int, isLoading, const char *, imageName),
+                 TP_FIELDS(ctf_integer(int, threadPtr, threadPtr)
+                           ctf_integer(int, isLoading, isLoading)
+                           ctf_string(imageName, imageName)))
+
+#endif /* _QT_TRACEPOINTS_H */
+
+#include <lttng/tracepoint-event.h>
diff --git c/src/tracesupport/tracesupport.pro i/src/tracesupport/tracesupport.pro
new file mode 100644
index 0000000..858f75d
--- /dev/null
+++ i/src/tracesupport/tracesupport.pro
@@ -0,0 +1,22 @@
+TARGET  = QtTraceSupport
+CONFIG += create_prl
+
+QT -= core gui
+
+load(qt_module)
+
+#################################
+
+DEFINES         += ENABLE_SA_TRACE
+CFLAGS          += -g3 -O1
+LIBS            += -llttng-ust -lurcu-bp -ldl
+
+
+#################################
+
+SOURCES = qt_tracepoints.c
+
+HEADERS += qt_tracepoints.h
+
+
+
diff --git c/src/widgets/graphicsview/qgraphicsscene.cpp i/src/widgets/graphicsview/qgraphicsscene.cpp
index bccdb1f..85c189e 100644
--- c/src/widgets/graphicsview/qgraphicsscene.cpp
+++ i/src/widgets/graphicsview/qgraphicsscene.cpp
@@ -251,6 +251,10 @@
 #include <private/qgesturemanager_p.h>
 #include <private/qpathclipper_p.h>
 
+#ifdef ENABLE_SA_TRACE
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 // #define GESTURE_DEBUG
 #ifndef GESTURE_DEBUG
 # define DEBUG if (0) qDebug
@@ -3280,6 +3284,10 @@ void QGraphicsScene::advance()
 */
 bool QGraphicsScene::event(QEvent *event)
 {
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtGraphicsSceneEvents, event->type());
+#endif // ENABLE_SA_TRACE
+
     Q_D(QGraphicsScene);
 
     switch (event->type()) {
diff --git c/src/widgets/kernel/qapplication.cpp i/src/widgets/kernel/qapplication.cpp
index 04d3982..e93b4eb 100644
--- c/src/widgets/kernel/qapplication.cpp
+++ i/src/widgets/kernel/qapplication.cpp
@@ -103,6 +103,11 @@ extern bool qt_wince_is_pocket_pc();  //qguifunctions_wince.cpp
 
 #include <qpa/qplatformwindow.h>
 
+#ifdef ENABLE_SA_TRACE
+    #define QT_TRACEPOINT_PER_LIBRARY_DEFINITION
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 //#define ALIEN_DEBUG
 
 static void initResources()
@@ -2480,6 +2485,10 @@ bool QApplicationPrivate::sendMouseEvent(QWidget *receiver, QMouseEvent *event,
     Q_ASSERT(nativeWidget);
     Q_ASSERT(buttonDown);
 
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtSendMouseEvent, event->buttons());
+#endif // ENABLE_SA_TRACE
+
     if (alienWidget && !isAlien(alienWidget))
         alienWidget = 0;
 
@@ -2955,6 +2964,13 @@ bool QApplication::notify(QObject *receiver, QEvent *e)
     case QEvent::KeyPress:
     case QEvent::KeyRelease:
         {
+#ifdef ENABLE_SA_TRACE
+            {
+                QKeyEvent* key = static_cast<QKeyEvent*>(e);
+                qt_tracepoint(Qt, qtNotifyKeyEvent, key->type(), key->key(), key->modifiers());
+            }
+#endif // ENABLE_SA_TRACE
+
             bool isWidget = receiver->isWidgetType();
             bool isGraphicsWidget = false;
 #ifndef QT_NO_GRAPHICSVIEW
@@ -3003,6 +3019,10 @@ bool QApplication::notify(QObject *receiver, QEvent *e)
 #endif
             }
             qt_in_tab_key_event = false;
+
+#ifdef ENABLE_SA_TRACE
+            qt_tracepoint(Qt, qtKeyEventConsumed, key->type(), key->key(), key->modifiers());
+#endif // ENABLE_SA_TRACE
         }
         break;
     case QEvent::MouseButtonPress:
@@ -3077,6 +3097,10 @@ bool QApplication::notify(QObject *receiver, QEvent *e)
 
             mouse->setAccepted(eventAccepted);
 
+#ifdef ENABLE_SA_TRACE
+            qt_tracepoint(Qt, qtMouseEventConsumed, mouse->buttons());
+#endif // ENABLE_SA_TRACE
+
             if (e->type() == QEvent::MouseMove) {
                 if (!pw)
                     break;
diff --git c/src/widgets/kernel/qwidgetbackingstore.cpp i/src/widgets/kernel/qwidgetbackingstore.cpp
index 9d024fd..3f09639 100644
--- c/src/widgets/kernel/qwidgetbackingstore.cpp
+++ i/src/widgets/kernel/qwidgetbackingstore.cpp
@@ -64,6 +64,10 @@
 #  include <qpa/qplatformnativeinterface.h>
 #endif
 
+#ifdef ENABLE_SA_TRACE
+    #include <QtTraceSupport/qt_tracepoints.h>
+#endif // ENABLE_SA_TRACE
+
 QT_BEGIN_NAMESPACE
 
 extern QRegion qt_dirtyRegion(QWidget *);
@@ -86,6 +90,10 @@ void QWidgetBackingStore::qt_flush(QWidget *widget, const QRegion &region, QBack
     Q_ASSERT(backingStore);
     Q_ASSERT(tlw);
 
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtFrameSync, 1);
+#endif // ENABLE_SA_TRACE
+
 #if !defined(QT_NO_PAINT_DEBUG)
     static int flushUpdate = qgetenv("QT_FLUSH_UPDATE").toInt();
     if (flushUpdate > 0)
@@ -121,6 +129,10 @@ void QWidgetBackingStore::qt_flush(QWidget *widget, const QRegion &region, QBack
     else
 #endif
         backingStore->flush(region, widget->windowHandle(), offset);
+
+#ifdef ENABLE_SA_TRACE
+    qt_tracepoint(Qt, qtFrameSync, 0);
+#endif // ENABLE_SA_TRACE
 }
 
 #ifndef QT_NO_PAINT_DEBUG
diff --git c/src/widgets/widgets.pro i/src/widgets/widgets.pro
index b12e283..183a03e 100644
--- c/src/widgets/widgets.pro
+++ i/src/widgets/widgets.pro
@@ -32,6 +32,7 @@ include(util/util.pri)
 include(statemachine/statemachine.pri)
 include(effects/effects.pri)
 
+contains(QT_CONFIG,tracesupport):CONFIG += tracesupport
 
 QMAKE_LIBS += $$QMAKE_LIBS_GUI
 
diff --git c/sync.profile i/sync.profile
index 690de60..cc2b9a9 100644
--- c/sync.profile
+++ i/sync.profile
@@ -16,6 +16,7 @@
     "QtANGLE/EGL" => "!$basedir/src/3rdparty/angle/include/EGL",
     "QtZlib" => "!$basedir/src/3rdparty/zlib",
     "QtOpenGLExtensions" => "$basedir/src/openglextensions",
+    "QtTraceSupport" => "$basedir/src/tracesupport",
 );
 %moduleheaders = ( # restrict the module headers to those found in relative path
 );
