From db75f6904227c0670c1a4a5e307b576d976a7739 Mon Sep 17 00:00:00 2001
From: Alexander Barabash <alexander_barabash@mentor.com>
Date: Tue, 6 Oct 2015 16:53:55 +0300
Subject: [PATCH 1/6] Add optional LTTng support in configure.

---
 configure.ac | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/configure.ac b/configure.ac
index df169de..f043afa 100644
--- a/configure.ac
+++ b/configure.ac
@@ -562,6 +562,26 @@ AC_CHECK_LIB(dl, dladdr,
    LIBS="$LIBS -ldl")
 CFLAGS="$save_cflags"
 
+dnl lttng tracepoints
+AC_ARG_ENABLE(lttng-tracepoints,
+   AC_HELP_STRING([--enable-lttng-tracepoints],
+                  [generate lttng tracepoints @<:@default=no@:>@]),
+   [AS_CASE(
+       [$enableval], [no], [], [yes], [], [auto], [],
+       [AC_MSG_ERROR([bad value "$enableval" for --enable-lttng-tracepoints])])],
+   [enable_lttng_tracepoints=no])
+AS_IF([test "x$enable_lttng_tracepoints" != "xno"],
+      [PKG_CHECK_MODULES([LTTNG_UST], [lttng-ust >= 2.0],
+                         [AC_DEFINE(GST_ENABLE_LTTNG_TRACEPOINTS, [1],
+                                    [Enable LTTng tracepoints.])
+                          LIBS="$LIBS $LTTNG_UST_LIBS -ldl"
+                          CFLAGS="$LIBS $LTTNG_UST_CFLAGS"
+                          enable_lttng_tracepoints=yes],
+                         [AS_IF([test "x$enable_lttng_tracepoints" = "xyes"],
+                                [AC_MSG_ERROR([lttng-ust not found])],
+                                [enable_lttng_tracepoints=no])])])
+AM_CONDITIONAL(GST_ENABLE_LTTNG_TRACEPOINTS, [test "x$enable_lttng_tracepoints" = "xyes"])
+
 dnl *** checks for dependency libraries ***
 
 dnl GLib
@@ -889,6 +909,7 @@ Configuration
 	Unit testing support       : ${BUILD_CHECK}
 
 	Debug                      : ${USE_DEBUG}
+	LTTng tracepoints          : ${enable_lttng_tracepoints}
 	Profiling                  : ${USE_PROFILING}
 
 	Building benchmarks        : ${BUILD_BENCHMARKS}
-- 
1.9.1

