From adf34d89c4a6f083e05e64e57988a536e0e08b6e Mon Sep 17 00:00:00 2001
From: Alexander Barabash <alexander_barabash@mentor.com>
Date: Tue, 8 Dec 2015 15:13:24 +0200
Subject: [PATCH 10/11] Added gst_latency_measurement_start/end/passed events.

---
 gst/gst_tracepoints.c |  1 +
 gst/gst_tracepoints.h | 15 +++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/gst/gst_tracepoints.c b/gst/gst_tracepoints.c
index aea8858..6c91a2f 100644
--- a/gst/gst_tracepoints.c
+++ b/gst/gst_tracepoints.c
@@ -33,6 +33,7 @@
 #include "gstevent.h"
 
 enum GstFlowTracepointType;
+static const gchar *gst_tracepoints_get_pad_element_name (GstPad * pad);
 static const gchar *gst_tracepoints_get_pad_element_name_if_needed (GstPad *
     pad, enum GstFlowTracepointType tracepoint_type);
 static guint16 gst_tracepoints_get_thread_id (void);
diff --git a/gst/gst_tracepoints.h b/gst/gst_tracepoints.h
index 6c6580f..103c45c 100644
--- a/gst/gst_tracepoints.h
+++ b/gst/gst_tracepoints.h
@@ -134,6 +134,21 @@ GST_TRACEPOINT_EVENT (gst_flow_event_caps,
                                  ctf_integer (unsigned long, flags, GST_CAPS_FLAGS (caps))
                                  ctf_string (text, str)))
 
+GST_TRACEPOINT_EVENT (gst_latency_measurement_start,
+                      TP_ARGS (GstPad *, pad, int, id),
+                      TP_FIELDS (ctf_integer (guint32, id, (guint32) (id))
+                                 ctf_string (element_name, gst_tracepoints_get_pad_element_name (pad))))
+
+GST_TRACEPOINT_EVENT (gst_latency_measurement_end,
+                      TP_ARGS (GstPad *, pad, int, id),
+                      TP_FIELDS (ctf_integer (guint32, id, (guint32) (id))
+                                 ctf_string (element_name, gst_tracepoints_get_pad_element_name (pad))))
+
+GST_TRACEPOINT_EVENT (gst_latency_measurement_passed,
+                      TP_ARGS (GstPad *, pad, int, id),
+                      TP_FIELDS (ctf_integer (guint32, id, (guint32) (id))
+                                 ctf_string (element_name, gst_tracepoints_get_pad_element_name (pad))))
+
 #endif /* _GST_TRACEPOINTS_H */
 
 #include <lttng/tracepoint-event.h>
-- 
2.2.1

